keylime/da/examples/sqldb.py

```diff
@@ -0,0 +1,135 @@
+    # -----------------------------------
+
+    def record_create(
+        self, agent_data, attestation_data, ima_policy_data=None, service="auto", signed_attributes="auto"
+    ):
+
+        # position the record in the git shadow. <localrepo>/<agentID>/<timestamp>
+        agentid = agent_data["agent_id"]
+        recordtime = datetime.datetime.now().strftime("%Y-%m-%d_%H:%M:%S")
+        recordtype = self._get_record_type(service)
+
+        # create the record, and sign it.
+        record_object = {}
+        self.record_signature_create(record_object, agent_data, attestation_data, service, signed_attributes)
+        rcrd = self._record_create(record_object, agent_data, attestation_data, ima_policy_data)
+
+        d = {"time": recordtime, "agentid": agentid, "record": rcrd}
+
+        try:
+            self.session.add((type2table(recordtype))(**d))  # pylint: disable=no-member
+            self.session.commit()  # pylint: disable=no-member
+        except Exception as e:
+            logger.error("Failed to create attestation record: %s", e)
+
+    # -----------------------------------
+    # create signature for an attestation record.
+    # -----------------------------------
+
+    def record_signature_create(
+        self, record_object, agent_data, attestation_data, service="auto", signed_attributes="auto"
+    ):
+        self._record_signature_create(record_object, agent_data, attestation_data, service, signed_attributes)
+        if "contents_file_path" in record_object and "signature_file_path" in record_object:
+            if self.tl_url.count("http") and self.tl_url.count("3000"):
+                getattr(importlib.import_module(self.st_imp_path + ".rekor"), "record_signature_create")(
+                    record_object, agent_data, self.tl_url, self.key_tls_pub, self.tmp_d_cl
+                )
+
+    # -----------------------------------
+    # retrieve a set of records
+    # -----------------------------------
+
+    def bulk_record_retrieval(self, record_identifier, record_list):
+        recordtype = self._get_record_type(self.svc)
+        tbl = type2table(recordtype)
+        for row in self.session.query(tbl).filter(tbl.agentid == record_identifier):  # pylint: disable=no-member
+            _decoded_record_object = self._record_decode(row.record)
+            self.record_signature_check(_decoded_record_object, record_identifier)
+            record_list.append(_decoded_record_object)
+
+    def build_key_list(self, agent_identifier, aik_list, service="auto"):
+        record_list = []
+        self.bulk_record_retrieval(agent_identifier, record_list)
+        self._build_key_list(record_list, aik_list)
+
+    def record_read(self, attestation_record_list, agent_identifier, service="auto"):
+        self.bulk_record_retrieval(agent_identifier, attestation_record_list)
+        self._record_read(attestation_record_list)
+
+    # -----------------------------------
+    # -----------------------------------
```

Comment:
drop?

Reply:
I missed a function header comment. I intensely dislike the python triple-quote way of commenting functions and I use these `#` based distancers instead. 

But I can step on my soul and rewrite all comments with triple quotes. I don't expect this to be urgent for the merge (pylint seems to be happy), but I don't want to impose my foibles on a public domain project either.

Reply:
> I intensely dislike the python triple-quote way of commenting functions

They are docs strings, and document the function (you can do later `help()` and inspect them). This is how Python works, so it is not much under debate.

```python
Python 3.10.7 (main, Sep 11 2022, 08:41:56) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def my_func():
...     """This resolve the Fermat problen in O(1)"""
...     pass
... 
>>> help(my_func)

Help on function my_func in module __main__:

my_func()
    This resolve the Fermat problen in O(1)
```

On the other side I think it is fair to no add doc strings to all the functions, so if you silently drop them I will not complain : )

Reply:
:-)